python_src/src/lamp_py/migrations/versions/performance_manager/017_37d97f420d54_add_prev_and_next_trip_stops.py

```diff
@@ -0,0 +1,178 @@
+            , vt.trip_id
+            , vt.vehicle_label
+            , vt.vehicle_consist
+            , vt.direction
+            , vt.direction_destination
+            , vt.static_trip_id_guess
+            , vt.static_start_time
+            , vt.static_stop_count
+            , vt.first_last_station_match
+            , vt.static_version_key
+            , vem.travel_time_seconds
+            , vem.dwell_time_seconds
+            , vem.headway_trunk_seconds
+            , vem.headway_branch_seconds
+            , COALESCE(vem.updated_on, ve.updated_on) as updated_on
+        FROM 
+            vehicle_events ve
+        LEFT JOIN
+            vehicle_trips vt
+        ON 
+            ve.trip_hash = vt.trip_hash
+        LEFT JOIN
+            vehicle_events prev_ve
+        ON
+            ve.pk_id = prev_ve.previous_trip_stop_pk_id
+        LEFT JOIN 
+            vehicle_event_metrics vem
+        ON
+            ve.trip_stop_hash = vem.trip_stop_hash
+        ;
+    """
+    op.execute("DROP VIEW IF EXISTS opmi_all_rt_fields_joined;")
+    op.execute(update_view)
+
+
+def downgrade() -> None:
+    update_view = """
+        CREATE OR REPLACE VIEW opmi_all_rt_fields_joined AS 
+        SELECT
+            vt.service_date
+            , ve.trip_hash
+            , ve.trip_stop_hash
+            , ve.stop_sequence
+            , ve.stop_id
+            , LAG (ve.stop_id, 1) OVER (PARTITION BY ve.trip_hash ORDER BY COALESCE(ve.vp_stop_timestamp,  ve.tu_stop_timestamp, ve.vp_move_timestamp)) as previous_stop_id
+            , ve.parent_station
+            , LAG (ve.parent_station, 1) OVER (PARTITION BY ve.trip_hash ORDER BY COALESCE(ve.vp_stop_timestamp,  ve.tu_stop_timestamp, ve.vp_move_timestamp)) as previous_parent_station
+            , ve.vp_move_timestamp
+            , COALESCE(ve.vp_stop_timestamp, ve.tu_stop_timestamp) as vp_tu_stop_timestamp
+            , vt.direction_id
+            , vt.route_id
+            , vt.branch_route_id
+            , vt.trunk_route_id
+            , vt.start_time
+            , vt.vehicle_id
+            , vt.stop_count
+            , vt.trip_id
+            , vt.vehicle_label
+            , vt.vehicle_consist
+            , vt.direction
+            , vt.direction_destination
+            , vt.static_trip_id_guess
+            , vt.static_start_time
+            , vt.static_stop_count
+            , vt.first_last_station_match
+            , vt.static_version_key
+            , vem.travel_time_seconds
+            , vem.dwell_time_seconds
+            , vem.headway_trunk_seconds
+            , vem.headway_branch_seconds
+            , COALESCE(vem.updated_on, ve.updated_on) as updated_on
+        FROM 
+            vehicle_events ve
+        LEFT JOIN
+            vehicle_trips vt
+        ON 
+            ve.trip_hash = vt.trip_hash
+        LEFT JOIN 
+            vehicle_event_metrics vem
+        ON
+            ve.trip_stop_hash = vem.trip_stop_hash
+        ;
+    """
```

Comment:
is there a tool or guide you use to help build this format this syntax?

i noticed when comparing this statement to the update_view string in the 16th revision that there was a small difference in format (an extra space in one of the coalesce statements).

Reply:
I don't have any specific guide that I follow, but wouldn't be opposed to adopting some more standard style guide for SQL.

I generally try to have one statement or expression per line, with helpful indenting. SQL is pretty flexible with white-space, in general. 

The extra space in the COALESCE statement wasn't on purpose. I sometimes copy and paste these SQL statements from a different editor where I confirmed the code was running, so things like that can get carried over.